[android] Adding pulsing circle to LocationComponent options

langsmith opened the pull request:
This pr adds a pulsing circle functionality to the Android `LocationComponent`. 

The following options/methods have been added to `LocationComponentOptions` for styling the pulsing circle:

- whether the pulsing circle is enabled or not
- circle's color
- circle's opacity
- number of milliseconds of length of a single pulse
- type of interpolated animation (linear, decelerate, accelerate, or bounce)
- whether the circle's opacity fades during the pulse


**_Fade enabled:_**
![ezgif com-resize 2](https://user-images.githubusercontent.com/4394910/52981124-85877680-3392-11e9-8a1b-672aa2086e14.gif)

_Fade disabled:_
![ezgif com-resize 3](https://user-images.githubusercontent.com/4394910/52981215-0fcfda80-3393-11e9-9f53-257a7ad427ac.gif)

langsmith commented:
Still definitely a work in progress but opening this pr to start discussion and review from @mapbox/maps-android .  Not sure whether my implementation is correct. Once things are 👌 on the back end, I can finish the test app example, write tests, finish Javadocs, etc.

Also, as you'll see in the _fade enabled_ GIF above, there's a weird "hitch" when a pulse is finished and a new one is about to begin. I assume it probably has to do with how I'm tracking things in the `PulsingLocationCircleAnimator` class. Any thoughts on a better way to achieve this effect? Something to do with the speed of the ValueAnimator and the Maps SDK's ability to catch up with new values?

langsmith commented:
I've got https://github.com/mapbox/mapbox-gl-native/pull/13942/files#diff-e6b14cce649c4c464b5c466cbda57938R76 , which is used [here](https://github.com/mapbox/mapbox-gl-native/pull/13942/files#diff-0c80662bdb24147525e13b17f29de6b4R163) and then ultimately [here](https://github.com/mapbox/mapbox-gl-native/pull/13942/files#diff-a30fc577bb1ed91a4c5e55630cbc8a46R1066). Doesn't seem to be working though. If you open the pulsing circle test app activity and then go back, it'll crash the app:

```
2019-02-18 16:11:09.093 29293-29293/? A/libc: Fatal signal 6 (SIGABRT), code -6 (SI_TKILL) in tid 29293 (pboxsdk.testapp), pid 29293 (pboxsdk.testapp)
```

Thoughts on how I might manage lifecycle with this background runnable and all?...

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -1056,6 +1061,10 @@ private void onLocationLayerStop() {
       locationEngine.removeLocationUpdates(currentLocationEngineListener);
     }
+    if (options.pulseEnabled()) {
+      Log.d(TAG, "onLocationLayerStop: about to stop animation ");
```
use Logger.java instead

langsmith replied:
✔️

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentOptions.java:
```diff
@@ -1041,6 +1223,18 @@ public LocationComponentOptions build() {
           + locationComponentOptions.elevation() + ". Must be >= 0");
       }
 
+      // TODO: Finish this logic below
```
implement todo

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationLayerController.java:
```diff
@@ -361,6 +375,21 @@ private void styleScaling(@NonNull LocationComponentOptions options) {
+  private void stylePulsingCircle(LocationComponentOptions options) {
+    if (mapboxMap.getStyle() != null) {
+      if (mapboxMap.getStyle().getLayer(PROPERTY_PULSING_CIRCLE_LAYER) != null) {
+        Log.d(TAG, "stylePulsingCircle: options.pulseEnabled() = " + options.pulseEnabled());
```
use Logger.java instead

langsmith replied:
✔️

langsmith commented:
>sub module pins were updated but not required for this PR?

Yes 😢

langsmith commented:
>might be good to refactor all the pulsing fields into a new object?

Yea, wasn't sure whether we should have completely separate builder or not for these pulsing options. I didn't make one as a start, 'cause it might be too many nested builders into one another. 3 levels deep: `PulsingCircleOptions` into `LocationComponentOptions` and then `LocationComponentOptions` [into  `LocationComponentActivationOptions`](https://github.com/mapbox/mapbox-gl-native/pull/13941/files#diff-2041cebcde0699dbbc13c3abd681e58a). Thoughts? @LukasPaczos ?

langsmith commented:
>Yea, wasn't sure whether we should have completely separate builder or not for these pulsing options

Note: @LukasPaczos followed up offline and confirmed that the current setup is fine. Pulsing options can be in `LocationComponentOptions` rather than being part of another builder process.

patrickkempff commented:
Is this ready to be merged?

langsmith commented:
@LukasPaczos , kept working on the refactoring to hook the pulsing into `MapboxAnimator.AnimationsValueChangeListener()`, etc. like we discussed a couple of weeks ago. The circle's expansion is still a little jittery and I'm not sure why. Would love another round of review from you and/or another chat to discuss. 

![ezgif com-resize (2)](https://user-images.githubusercontent.com/4394910/59953898-118bd680-9437-11e9-89f5-6a4e60c48260.gif)

![ezgif com-resize (3)](https://user-images.githubusercontent.com/4394910/59953897-118bd680-9437-11e9-8b6f-658ba1e6da7f.gif)

langsmith commented:
hey @patrickkempff , thanks for your interest. It's not quite ready, but as evidenced by my comment above, I'm getting closer.

patrickkempff commented:
@langsmith I see. Keep up the good work! I am looking forward for this 👍

langsmith commented:
Slowly continuing work on this. It's looking good and the pulsing code is correctly set up to match our other Animators. @LukasPaczos and I aren't sure why the pulsing doesn't resume when the activity is returned to or when a map style is changed. Need to look into what's going with that.


![ezgif com-resize (1)](https://user-images.githubusercontent.com/4394910/61322156-ea1ff380-a7c1-11e9-9e75-994c66b7efe2.gif)

langsmith commented:
Status update. Things are looking great. I've added a number of different tests but I'm wrestling with a weird testing bug that's happening during the building of `LocationComponentOptions`.

```
  java.lang.IllegalArgumentException: You cannot set both layerAbove and layerBelow options. Choose one or the other.
        at com.mapbox.mapboxsdk.location.LocationComponentOptions$Builder.build(LocationComponentOptions.java:1300)
        at com.mapbox.mapboxsdk.location.LocationComponentOptions.createFromAttributes(LocationComponentOptions.java:373)
        at com.mapbox.mapboxsdk.location.LocationComponentOptions.builder(LocationComponentOptions.java:398)
        at com.mapbox.mapboxsdk.location.LocationAnimatorCoordinatorTest.setUp(LocationAnimatorCoordinatorTest.kt:42)
``` 

![Screen Shot 2019-08-02 at 4 47 05 PM](https://user-images.githubusercontent.com/4394910/62403875-2e8eeb80-b545-11e9-99c3-951885478c7d.png)




The pulsing now continues after returning to the activity from a different app. Pulsing now stops when the `LocationComponent` is deactivated and the pulsing resumes when the `LocationComponent` is reactivated. Pulsing also resumes when the map style has changed.

![ezgif com-resize (1)](https://user-images.githubusercontent.com/4394910/62000544-39053d00-b08e-11e9-9a0b-a268b79ad8d2.gif)

![ezgif com-resize (2)](https://user-images.githubusercontent.com/4394910/62000562-a2854b80-b08e-11e9-8d52-39c88f4fd28b.gif)

`cancelPulsingCircleAnimation()` is now working. It cancels the pulsing animator and hides the pulsing circle's layer.
![ezgif com-resize (3)](https://user-images.githubusercontent.com/4394910/62000569-f4c66c80-b08e-11e9-981b-617db69ab0b1.gif)

Along with `CustomizedLocationPulsingCircleActivity`, which has the speed/color toggle, I've added `BasicLocationPulsingCircleActivity`. It shows the basic default pulsing circle:

![ezgif com-resize](https://user-images.githubusercontent.com/4394910/62000504-8a60fc80-b08d-11e9-93ad-cee77e130c0e.gif)

langsmith commented:
Ok @LukasPaczos , things are looking good visually and all tests are passing. This is ready for yet another round of review when you have the time and are in the mood.

patrickkempff commented:
Looks good! Looking forward to use this. Any updates when this PR will be merged? 👍

tobrun commented:
We are removing platform code from gl-native in https://github.com/mapbox/mapbox-gl-native/pull/15970. If you want to see this PR merged, it will require re-implementation in https://github.com/mapbox/mapbox-gl-native-android.

DntPullALockett commented:
@malwoodsantoro  When can we get this re implemented?

langsmith commented:
For anyone following this pr and/or interested in using this feature, you should know that this feature has been added to the Mapbox Maps SDK for Android at https://github.com/mapbox/mapbox-gl-native-android/pull/172. 

The current Mapbox Maps SDK for Android is in the https://github.com/mapbox/mapbox-gl-native-android repository because the Mapbox team decided to move the Maps SDK for Android to a standalone repo: https://github.com/mapbox/mapbox-gl-native/issues/15971

Cloud9Clone commented:
When the new version 9.2.0. will be officially announced? I see that on the Mapbox page stays still 9.1.0. ?

langsmith commented:
Hey @Cloud9Clone, `9.2.0` is out and available. It went out yesterday https://github.com/mapbox/mapbox-gl-native-android/releases/tag/android-v9.2.0 

https://docs.mapbox.com/android/maps/overview/ and other docs pages just haven't been updated yet. 

Docs about pulsing `LocationComponent` are at the bottom of the https://docs.mapbox.com/android/maps/overview/location-component/#active-styling-options section.
